src/mixing.html

```diff
@@ -15,53 +15,78 @@ <h2 id="mixing">Annotating MathML</h2>
+  the first child of the 3rd child of the element. An argumnt
+  reference such as <code>$name</code> refereces a descendent element
+  that has an attriute <code>arg="name"</code>. unlike <code class="attribute">id</code>
+  attributes, <code class="attribute">arg</code> do not have to be
+  unique within a document, when searching for a matching element the
+  search should only consider descendns and not consider descendents
+  of any element that has an <code class="attribute">intent</code> or
+  <code class="attribute">arg</code> attribute.</li>
+  <li><em>literal numbers</em> such as <code>2.5</code> denote
+  themselves. <span class="ednote">0.25e1?</span></li>
+  <li><em>literal identifiers</em> Non numeric literals reference
+  terms in the intent dictionary, currently located at <a
+  href="https://docs.google.com/spreadsheets/d/1EsWou1K5nxBdLPvQapdoA9h-s8lg_qjn8fJH64g9izQ/edit#gid=1358098730">Google
+  Sheets</a>. Note the list of [=intent=] identifiers is
+  <em>Open</em>: level 1 is a curated list maintained by the Math Working
+  Group and should be implemented by any implemenation of MathML
+  intent, so that suitable audio renderings of the level 1 terms will
+  be generated. Terms in other levels may be contributed; the
+  dictionary may give speech hints, but a system may also fall back on
+  reading the identifier name as given.</li>
+  <li><em>the wildcard</em>, <code>*</code> denotes all the children
+  <span class="ednote">just mi and mn children?</span> of the element so
+  that <code>intent="somefunction(*)"</code> is equivalent to
+  <code>intent="somefunction($1,$2, ...)"</code></li>
+  </ul>
+
+<p>The key component here is the literal; they are intended to correspond to some mathematical concept or operator, or some application specific quantity or operation; that is it represents some “meaning”. Ultimately, the goal is to translate this virtual content tree into text or speech (in different languages), braille or some other form. The preferred translation may also depend on context and user profile. On the one hand, the most natural translation of a given expression can depend on the operator. It is thus desirable to have a set of common known meaning with translation rules. Thus power(x,2) might be rendered as “the 2nd power of x” or “x squared”, depending on the user and agent.</p>
+
   <section>
    <h4 id="mixing_intent_grammar">The Grammar for <code class="attribue">intent</code></h4>
 
    <pre>
-intent := literal | selector | intent '(' wildcard ')' | intent '(' intent [ ',' intent ]* ')'
-literal      := [letters|digits|_|-]*
-selector     := ( argref | numref ) [ '/' selector ]?
-argref       := '$' NCName
-numref       := '$' [digit]+
-wildcard     := '@'
+intent   := literal | selector | intent '(' wildcard ')' | intent '(' intent [ ',' intent ]* ')'
+literal  := ( letters | digits | '.' | '_' | '-' )*
+selector := argref | numref
+argref   := '$' NCName
+numref   := '$' [digit]+ ( '/' numref )?
```

Comment:
First, isn't `numref` actually a `pathref` given the slashes?

For the record I'm still not fond of including numbers or paths given that `argref` is more human-friendly and covers an equivalent class of expressions. Having two separate referencing schemes is a bit overkill for a simple mini-language.

I probably don't have enough group support to stop including both, but if I did - I would.

Reply:
pathref yes perhaps. It's only a name.

Actually that was a very late change just before I pushed as I didn't want to have to document what `$foo/$3/$bar` (allowed in the xisting gh-pages draft) means so I decided to disallow it.

I'm also not totally convinced by the need to have `/` (or `*`)  but I'd rather leave them in for now until we have some words and/or implementation around default values for `intent` If we end up not needing this to describe `intent` then we can probably drop it.  

Using `$argref` can make the same references but  requires modifying the child elements which may not be so convenient. That is, when generating a non-fraction mfrac you may be able to generate `mfrac intent="fooobar($1,$2)"` then forget about it, if you generate `mfrac intent="fooobar($a,$b)"` then your generator needs to recursively "remember" to add `arg="a"` and `arg="b"` in the right places. So it  probably ends up being simpler to specify default intents using a path. If you are adding intents by hand to an existing tree, using argref is simpler I agree.

Reply:
> If we end up not needing this to describe intent then we can probably drop it.

For the record, *I* currently don't need this to describe intent, and would be quite OK with dropping it. I'm really not fond of preemptively adding untested bits to the grammar, *given that* we have tested a wide range of expression with a more constrained syntax.

Just as one example of unintended side-effects, the grammar above now allows the following as a valid intent value:
```
.(.. , ..., , , ,)
```

To my taste, the literal rule is along the lines of:
```
literal  := letter+ ('-' letter+)*
```
focusing on the encyclopedic intent setup, where we have ample examples for the concrete needs of AT. The numeric annotations strike me as somehow off-topic until we get some examples where they appear to be a must-have.

Reply:
> > If we end up not needing this to describe intent then we can probably drop it.
> 
> For the record, _I_ currently don't need this to describe intent, and would be quite OK with dropping it. I'm really not fond of preemptively adding untested bits to the grammar, _given that_ we have tested a wide range of expression with a more constrained syntax.

well that's rather the point, you, Sam, Bruce, Neil, then I all had different syntaxes in intent suggestions.
We need to converge on a single syntax.

> 
> Just as one example of unintended side-effects, the grammar above now allows the following as a valid intent value:
> 
> ```
> .(.. , ..., , , ,)
> ```

actually that was an _intended_  feature, the grammar I originally had in the unified intent issue did not allow empty arguments in the intent function syntax, just allowed a complely empty intent. @brucemiller  commented that his grammar had allowed, and used empty argument slots. The grammar only needs to constrain so far, it will naturally allow some non sensical expressions which need to be handled at a higher level. We could change some `*` to `+` if empty slots turn out  not to be needed, but we'll see,. It would of course be possible to have a grammar that allowed empty argument slots but forced a non empty identifier in the function name, but if we want to allow empty anyway keping the grammar simple seemed preferable rather than forcing special non-empty/empty productions just for this. So I allowed empty everywhere in this draft.

> 
> To my taste, the literal rule is along the lines of:
> 
> ```
> literal  := letter+ ('-' letter+)*
> ```
> 

see above...

> focusing on the encyclopedic intent setup, where we have ample examples for the concrete needs of AT. The numeric annotations strike me as somehow off-topic until we get some examples where they appear to be a must-have.

we had a general aim that people should be able to do these attributions at different levels of granularity. If you want to go down to the leaves and have  `<mn>1</mn>` then some designs would advocate always having an explicit possibility for stating the default and that would seem to be `<mn intent="2">2</mn>`  More usefully if you want to annotate larger expressions without annotating the whole tree you might want to end up with  `<math intent="plus(power(cos(x),2),power(sin(x),2))"`  (by the way if we are aiming for accesibility rather than computation, why don't we allow infix functions in intent?, I want to write  `intent="power(sin(x),2)+power(cos(x),2) = 1" but in any case these seem to use numberic literals in intent.

The semanitics-mini paper had `power(x,2)` in is opening examples, which used 2....

Reply:
>  it will naturally allow some non sensical expressions which need to be handled at a higher level.

Well, the nonsensical uses of dot and underscore are very easy to prevent in the grammar itself, similarly to how i guarded the dash in my comment above. The higher-level problem is that if we add constructs without double and triple-checking their utility, we end up with a messier outcome that has a greater learning curve for less clarity.

An empty argument pattern is just ` '(' ')'` which can be allowed standalone, if need be. Though here again I'm interested in what AT purpose that serves, since it looks suspiciously close to a programming function call, rather than an AT aid.

> an explicit possibility for stating the default

My worry is that this line of reasoning eventually implies full Unicode support in literals, not just numbers. See the start of [arXiv:1810.00769](https://ar5iv.labs.arxiv.org/html/1810.00769) for one example:
```xml
<mi intent="♣">♣</mi>
<mi intent="♡">♡</mi>
```
People make use of a lot of characters beyond letters and digits in their MathML token elements, especially `mi` and `mo` nodes, as I'm sure you know. Stating the default intent to be "as written", requires us to allow hundreds, if not thousands, of characters in the `literal` rule. On top of that, having a grasp of compound objects, such as numbers in scientific notation, requires additionally growing the grammar if we wanted to validate them.

From where I'm standing, none of this extra boilerplate is needed. If one had to, they could use an arg reference in the complex expressions, as in `power(speed-of-light, $2)`. And, most productively, we can completely avoid annotating nodes that follow the specified defaults. `<mn>2</mn>` should never need a redundant `intent="2"`.

As another alternative, XPath has the single `text()` selector for text content. Allowing that one special keyword exception appears to me as more manageable than opening the door to all of Unicode, and/or growing the grammar complexity.

---

The challenging example I initially was fond of was `second-derivative`, where "the second derivative of f" could be written:
```xml
<msup>
  <mi>f</mi>
  <mo>″</mo>
</msup>
```
One way to mark that up would be `derivative(f,2)`, where the `2` has no literal equivalent in the MathML, and would require numbers added to the `literal` rule. 

But there is another way - using a [second-derivative](https://en.wikipedia.org/wiki/Second_derivative) literal, as in `second-derivative(f)`. Having an encyclopedic page for that string naturally shifts my own preference towards using that keyword, with no argument beyond the f (or a `$ref` to f), instead of an open-ended numeric argument.

Reply:
If we want to have an open-ended components (ie. not only keywords specified in some to-be-determined dictionary), then we will have to allow (most of) Unicode.  I would have been inclined to allow literals to include *all* characters not explicitly given some specific grammar here (parens, comma, $), but that severely limits any future expansion.  Slightly safer approach would reserve a set of punctuating chars for potential future use.

Explicitly marking arbitrary literals using `text()`, or quotes (embedded, argh!) may be safer.  But that opens up more potential for user error. And also seems to introduce a distinction between open ended text and keywords (That could be argued as either good or bad).

Reply:
@brucemiller  yes I agree that we will need most of unicode, taking NCName might be a coherent choice as that avoids most punctuation and has the advantage of being someone else's arbitrary choice. Otherwise I was assuming that `letter` in the grammar meant `[:letter:]` ie whatever Unicode says is a letter, not `[a-z]`

Reply:
My working assumption for letters has been exactly "whatever Unicode says is a letter", since a main hope for the literals is to represent "concept names". And names are comprised of alphabetic letters, in whichever language an author is happening to write them down.

But I also never wanted to be providing tooling for literal content *other than* concept names. A very minimal and constrained initial scope of intent.